src/unix/lwt_unix_unix.h

```diff
@@ -1860,7 +1851,7 @@ static value result_readdir_n(struct job_readdir_n *job)
         result = caml_alloc(job->count, 0);
         long i;
         for (i = 0; i < job->count; i++) {
-            Store_field(result, i, caml_copy_string(job->entries[i].d_name));
+            Store_field(result, i, caml_copy_string(job->entries[i]));
```

Comment:
IIRC: the code before this PR allocated a flat array of `dirent`s, i.e. if it calculated that a `dirent` was 1024 bytes and needed to allocate 32 of them, it would allocate 32KB of space in one block. The strings were written directly into various parts of this single block. We controlled the maximum length of these strings when calling `readdir_r`.

The new code inserts a level of indirection, because we don't know how long the strings returned by `readdir` will be. So, for reading 32 entries, it is now allocating space for 32 8-byte pointers in one block, and each one of those pointers may then be set to the address of a separately-allocated string (by `strdup`). We need to free these separate string blocks.